### PR TITLE
OpenFilterPanel

### PR DESCRIPTION
**Pull request by shaobinspire**

### CODE REVIEW

**Comment by kranar** [comment]
Lot of good work, looks almost done but some minor changes need to be made.

File: Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp [hunk 1]
```diff
@@ -101,6 +101,9 @@ namespace Spire {
   template<typename T>
   QValidator::State AnyValueModel::WrapperValueModel<T>::set(
       const Type& value) {
+    if(!value.has_value()) {
```
Comment (kranar) [inline Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp:104]: Undo this change.

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+    style.get((Hover() || Press()) / Body() >> is_a<Icon>()).
+      set(BackgroundColor(QColor(Qt::transparent)));
+    style.get(Disabled() / Body() >> is_a<Icon>()).
+      set(Visibility::NONE);
+    return style;
+  }
+
+  auto LIST_ITEM_STYLE() {
+    auto style = StyleSheet();
+    style.get((Any() || Hover() || Press() || Focus() || Selected())).
```
Comment (kranar) [inline Applications/Spire/Source/Ui/OpenFilterPanel.cpp:34]: Remove extraneous parenthesis.

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+      set(BackgroundColor(QColor(0xFFFFFF))).
+      set(border_size(0)).
+      set(PaddingRight(scale_width(10)));
+    return style;
+  }
+
+  auto display_text(OpenFilterPanel::Mode mode) {
+    if(mode == OpenFilterPanel::Mode::INCLUDE) {
+      static const auto value = QObject::tr("Include");
+      return value;
```
Comment (kranar) [inline Applications/Spire/Source/Ui/OpenFilterPanel.cpp:44]: Avoid making copies.

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+    } else {
+      static const auto value = QObject::tr("Exclude");
+      return value;
+    }
+  }
+}
+
+class DeletableItem : public QWidget {
+  public:
+    using DeleteSignal = Signal<void()>;
```
Comment (kranar) [inline Applications/Spire/Source/Ui/OpenFilterPanel.cpp:54]: Space after ```void```.

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+    }
+  }
+
+  std::any parse(const QString& query) override {
+    auto value = m_source->parse(query);
+    if(!value.has_value()) {
+      return value;
+    }
+    if(m_matches_set.contains(displayTextAny(value))) {
+      static auto result = std::any();
```
Comment (kranar) [inline Applications/Spire/Source/Ui/OpenFilterPanel.cpp:106]: What is the purpose of this ```static``` variable?

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+  QtPromise<std::vector<std::any>> submit(const QString& query) override {
+    return m_source->submit(query).then([=] (auto&& source_result) {
+      auto result = [&] {
+        try {
+          return source_result.Get();
+        } catch(const std::exception&) {
+          return std::vector<std::any>();
+        }
+      }();
+      std::erase_if(result, [=] (const auto& value) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/OpenFilterPanel.cpp:121]: Incorrect capture.

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+
+    void on_update(Mode mode, bool value) {
+      get_button(mode)->get_current()->set(value);
+      if(value) {
+        auto blocker = shared_connection_block(m_connection);
+        m_model->set(mode);
+      }
+    }
+
+    void on_model_update(const Mode mode) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/OpenFilterPanel.cpp:181]: Don't pass by ```const``` value.

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp [hunk 8]
```diff
@@ -0,0 +1,377 @@
...
+  return item;
+}
+
+void OpenFilterPanel::on_input_box_submission(const AnyRef& submission) {
+  m_input_box->get_current()->set({});
+  m_matches->push(submission);
+}
+
+void OpenFilterPanel::on_list_view_current(const optional<int>& current) {
+  m_list_view->setFocusPolicy(Qt::NoFocus);
```
Comment (kranar) [inline Applications/Spire/Source/Ui/OpenFilterPanel.cpp:338]: Why is the list view's focus policy constantly being set?

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp [hunk 9]
```diff
@@ -0,0 +1,377 @@
...
+void OpenFilterPanel::on_mode_current(Mode mode) {
+  submit();
+}
+
+void OpenFilterPanel::on_matches_operation(
+    const AnyListModel::Operation& operation) {
+  visit(operation,
+    [&] (const AnyListModel::AddOperation& operation) {
+      auto item = m_list_view->get_list_item(operation.m_index);
+      set_style(*item, LIST_ITEM_STYLE());
```
Comment (kranar) [inline Applications/Spire/Source/Ui/OpenFilterPanel.cpp:350]: Why is the style being set here on every add operation?

File: Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp (hunk 1 above)
Comment (shaobinspire) [reply Applications/Spire/Include/Spire/Spire/AnyValueModel.hpp:104]: The intention is to clear the value of the `AnyValueModel` from the `OpenFilterPanel`. The `OpenFilterPanel` knows that the type of the current of the `AnyInputBox` is `AnyRef`, and it can set an empty `AnyRef` to the `current` to clear the `AnyInputBox`. If this code gets removed, `any_cast` will return a invalid pointer because the type of the empty `AnyRef` is void but the casted type is `std::any` which is the type of the `current` of the `ComboBox`, and there is no way to clear the `AnyInputBox` from the `OpenFilterPanel`.

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp (hunk 9 above)
Comment (shaobinspire) [reply Applications/Spire/Source/Ui/OpenFilterPanel.cpp:350]: I tried to set the style of the `ListItem` in the `ListView`, but the style couldn’t be applied to the newly added `ListItem` through the `ListView`.

Comment (shaobinspire) [reply Applications/Spire/Source/Ui/OpenFilterPanel.cpp:350]: I found a solution to solve it.

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+
+void OpenFilterPanel::on_mode_current(Mode mode) {
+  submit();
+}
+
+void OpenFilterPanel::on_matches_operation(
+    const AnyListModel::Operation& operation) {
+  visit(operation,
+    [&] (const AnyListModel::AddOperation& operation) {
+      update_style(*m_list_view, [] (auto&) {});
```
Comment (kranar) [inline Applications/Spire/Source/Ui/OpenFilterPanel.cpp:333]: What does this do?

File: Applications/Spire/Source/Ui/ScrollBar.cpp [hunk 11]
```diff
@@ -236,6 +236,7 @@ void ScrollBar::update_thumb() {
       style.get(Any()).set(PaddingLeft(m_thumb_position));
     }
   });
+  update();
```
Comment (kranar) [inline Applications/Spire/Source/Ui/ScrollBar.cpp:239]: What does this do?

**Review by kranar — CHANGES_REQUESTED** [review]
Major improvement.

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp (hunk 10 above)
Comment (shaobinspire) [reply Applications/Spire/Source/Ui/OpenFilterPanel.cpp:333]: When a ListItem is added to the ListView, the newly added ListItem cannot be styled, so I have to style the ListItem or update the style of the ListView when the add operation happens.

Comment (kranar) [reply Applications/Spire/Source/Ui/OpenFilterPanel.cpp:333]: That's good, when you come across bugs in the future, instead of trying to work around them, report them in Asana so they can get fixed.

File: Applications/Spire/Source/Ui/ScrollBar.cpp (hunk 11 above)
Comment (shaobinspire) [reply Applications/Spire/Source/Ui/ScrollBar.cpp:239]: Here I am trying to solve the bug of the ScrollBar. When enough ListItems are added into the ListView and the vertical scroll bar becomes visible, the thumb is not draggable even though the ListItems can move with the mouse dragging.

I will remove it and report the issue in Asana.

Comment (kranar) [reply Applications/Spire/Source/Ui/ScrollBar.cpp:239]: You can keep this in here, just wasn't sure what it was doing.

File: Applications/Spire/Source/Ui/OpenFilterPanel.cpp (hunk 10 above)
Comment (shaobinspire) [reply Applications/Spire/Source/Ui/OpenFilterPanel.cpp:333]: For this one, should I remove it or keep it?

Comment (kranar) [reply Applications/Spire/Source/Ui/OpenFilterPanel.cpp:333]: You can keep it, but I need to fix this bug, so in the future just find a way to make it clear that this is working around a bug, preferably by creating a task in Asana. In your Asana task you can refer to this line (along with the commit) and point out that this is a workaround to a bug.

**Review by kranar — APPROVED** [review]
